### PR TITLE
Integrated tsx-editor into example card and fabric website

### PR DESCRIPTION
**Pull request by nishikaza**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

Work in progress. Editor component now slowly being integrated into fabric website workflow. Added state management with ExamplesSection to manage active example card and ensure only one editor is active at once. Changed exports in tsx-editor and updated dependencies in a few places to get to working state.

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9769)

### CODE REVIEW

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.types.ts [hunk 1]
```diff
@@ -27,6 +27,12 @@ export interface IExampleCardProps {
 
   /** Optional override styles */
   styles?: IStyleFunctionOrObject<IExampleCardStyleProps, IExampleCardStyles>;
+
+  /** On click handler to ensure only one code editor instance is shown at once */
+  onClick?: (card: string) => void;
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.types.ts]: This should have a more specific name, like `onShowHideCodeClick`. And maybe an additional boolean parameter to indicate if it's being shown or hidden.

Comment (ecraig12345) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.types.ts]: Or instead of adding a boolean parameter, you could add documentation indicating that if the code is being hidden, pass undefined/an empty string.

Comment (nishikaza) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.types.ts]: @ecraig12345 Was so caught up in switching between different editors that I didn't even realize basic toggle hide didn't work, pushed fix :)

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 2]
```diff
@@ -172,9 +172,15 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
   };
 
   private _onToggleCodeClick = () => {
-    this.setState({
-      isCodeVisible: !this.state.isCodeVisible
-    });
+    if (this.props.isCodeVisible) {
+      this.props.onClick!('');
```
Comment (cliffkoh) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: can we please avoid the use of `!` postfix operator, or if deemed to make a lot of sense, include a JSDoc explaining why? As more and more `!` creeps into the code (and the moment you see one, others will be tempted to add more), the harder it gets to reason over correctness and type safety significantly degrades...

File: apps/fabric-website/webpack.serve.config.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 let path = require('path');
```
Comment (NikoAri) [inline apps/fabric-website/webpack.serve.config.js]: >let [](start = 0, length = 3)

Just curious, should this file name be "webpack.server.config.js"? Or "webpack.serve.config.js" is correct?

File: change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "comment": "Website changes to work with monaco editor state management",
+  "type": "minor",
+  "packageName": "@uifabric/example-app-base",
+  "email": "nishikaza@berkeley.edu",
```
Comment (NikoAri) [inline change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json]: >nishikaza@berkeley.edu [](start = 12, length = 22)

use your ms alias please

File: change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "comment": "Website changes to work with monaco editor state management",
```
Comment (NikoAri) [inline change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json]: >monaco [](start = 43, length = 6)

(nit) capitalize

File: change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "comment": "Website changes to work with monaco editor state management",
+  "type": "minor",
+  "packageName": "@uifabric/fabric-website",
+  "email": "nishikaza@berkeley.edu",
```
Comment (NikoAri) [inline change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json]: >nishikaza [](start = 12, length = 9)

use ms alias please

File: packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx [hunk 7]
```diff
@@ -27,7 +32,9 @@ export const ExamplesSection: React.StatelessComponent<IExamplesSectionProps> =
             const { view, ...exampleProps } = example;
             return (
               <div key={example.title + '-key'} className={styles.subSection}>
-                <ExampleCard {...exampleProps}>{view}</ExampleCard>
+                <ExampleCard {...exampleProps} onClick={handleClick} isCodeVisible={exampleProps.title === activeEditor}>
```
Comment (NikoAri) [inline packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx]: >activeEditor [](start = 107, length = 12)

(nit) activeEditorTitle?

Comment (NikoAri) [inline packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx]: >exampleProps.title [](start = 84, length = 18)

It's a little strange that we need to compare strings of all editors to figure out active editor. It should be some boolean flag inside of exampleProps, that specifies which editor is active.

File: change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json (hunk 6 above)
Comment (nishikaza) [reply change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json]: I think this was just a change file that was auto-generated, I updated my primary email in Github for future commits

---
In reply to: [302628097](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302628097) [](ancestors = 302628097)

File: change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json (hunk 5 above)
Comment (nishikaza) [reply change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json]: Came from commit message, will capitalize in future commits

---
In reply to: [302627938](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302627938) [](ancestors = 302627938)

File: change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json (hunk 4 above)
Comment (nishikaza) [reply change/@uifabric-example-app-base-2019-07-10-17-15-36-monaco-editor.json]: Auto-generated from change file, updated primary email in Github for future commits

---
In reply to: [302627750](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302627750) [](ancestors = 302627750)

File: apps/fabric-website/webpack.serve.config.js (hunk 3 above)
Comment (nishikaza) [reply apps/fabric-website/webpack.serve.config.js]: "webpack.serve.config.js" is correct here

---
In reply to: [302627456](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302627456) [](ancestors = 302627456)

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 8]
```diff
@@ -80,8 +78,10 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
             </div>
           );
 
+          const editor = <Editor onChange={this._editorOnChange} width={800} height={500} code={code!} language="typescript" />;
```
[the remark this answers is not in the document]
Comment (nishikaza) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: These values originally came from prototyping, changed width to autofit and kept width static at 500 since autofitting height results in an incredibly small editor

---
In reply to: [302628550](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302628550) [](ancestors = 302628550)

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx (hunk 2 above)
Comment (nishikaza) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Added extra conditional to check and avoid '!' operator

---
In reply to: [302383654](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302383654) [](ancestors = 302383654)

File: change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json (hunk 6 above)
Comment (khmakoto) [reply change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json]: Just so you know, even though these were auto-generated, once they are generated you can modify them under the `change` folder in your project, so you could potentially address these comments if you want 😄

File: packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx (hunk 7 above)
Comment (nishikaza) [reply packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx]: This check is effectively passing in a boolean isCodeVisible as a prop for each ExampleCard so it accomplishes the same goal. The check needs to be done here in ExamplesSection though to pass the property down to each ExampleCard.

---
In reply to: [302632279](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302632279) [](ancestors = 302632279)

File: change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json (hunk 6 above)
Comment (nishikaza) [reply change/@uifabric-fabric-website-2019-07-10-17-15-36-monaco-editor.json]: Updated :)

---
In reply to: [302672819](https://github.com/OfficeDev/office-ui-fabric-react/pull/9769#discussion_r302672819) [](ancestors = 302672819)

File: apps/fabric-website/webpack.serve.config.js (hunk 3 above)
Comment (ecraig12345) [reply apps/fabric-website/webpack.serve.config.js]: "serve" is standard in our projects as well as various other open source projects.

File: packages/codepen-loader/just.config.js [hunk 9]
```diff
@@ -2,17 +2,7 @@
...
-  task,
-  series,
-  parallel,
-  option,
-  condition,
-  tscTask,
-  copyTask,
-  argv
-} = require('just-scripts');
+const { taskPresets, task, series, parallel, option, condition, tscTask, copyTask, argv } = require('just-scripts');
```
Comment (ecraig12345) [inline packages/codepen-loader/just.config.js]: Revert the changes to this file

File: packages/example-app-base/package.json [hunk 10]
```diff
@@ -44,10 +44,12 @@
   },
   "dependencies": {
     "@microsoft/load-themed-styles": "^1.7.13",
-    "@uifabric/fluent-theme": "^7.0.4",
-    "@uifabric/set-version": "^7.0.1",
+    "@uifabric/fluent-theme": "^7.0.3",
```
Comment (ecraig12345) [inline packages/example-app-base/package.json]: Revert the fluent-theme and set-version versions

File: apps/fabric-website/index.html [hunk 11]
```diff
@@ -206,5 +206,7 @@
         return decodeURIComponent(results[2].replace(/\+/g, ' '));
       }
     </script>
+    <script src="//unpkg.com/react-dom@16/umd/react-dom.development.js"></script>
```
Comment (ecraig12345) [inline apps/fabric-website/index.html]: You shouldn't need this part

File: packages/tsx-editor/src/components/Editor.tsx [hunk 12]
```diff
@@ -23,10 +17,22 @@ export const Editor: React.FunctionComponent<IEditorProps> = (props: IEditorProp
       experimentalDecorators: true,
       preserveConstEnums: true,
       outDir: 'lib',
-      module: monaco.languages.typescript.ModuleKind.CommonJS,
+      module: monaco.languages.typescript.ModuleKind.ESNext,
       lib: ['es5', 'dom']
     });
 
+    monaco.languages.typescript.typescriptDefaults.setDiagnosticsOptions({ noSemanticValidation: true });
```
Comment (ecraig12345) [inline packages/tsx-editor/src/components/Editor.tsx]: What does this do?

File: packages/tsx-editor/src/components/EditorContext.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import * as monaco from 'monaco-editor';
+import { ITextModel } from './Editor.types';
+
+interface IProviderProps {
+  code: string;
+}
+
+const defaultValues = {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/components/EditorContext.tsx]: Defaults of undefined might be more appropriate here.

File: packages/tsx-editor/src/components/EditorContext.tsx [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    return;
+  },
+  setID: (newID: string) => {
+    return;
+  }
+};
+
+export const EditorContext = React.createContext(defaultValues);
+
+export class EditorProvider extends React.Component<IProviderProps> {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/components/EditorContext.tsx]: What's the type of this component's state?

File: packages/tsx-editor/src/components/EditorPreview.tsx [hunk 15]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+import { Label } from 'office-ui-fabric-react';
+interface IEditorPreview {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/components/EditorPreview.tsx]: This needs to be exported and should be called IEditorPreviewProps (also please add a line before the declaration)

File: packages/tsx-editor/src/components/EditorPreview.tsx [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  id: string;
+  error?: string;
+}
+
+const errorStyle = {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/components/EditorPreview.tsx]: You should probably give this a type (I think it might be React.CSSProperties?)

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
+export interface ITransformedExample {
+  output?: string;
+  error?: string;
+}
+
+export function transformExample(example: string, id: string) {
+  const classNamePattern = new RegExp('(?<=var )(.*)(?= = /\\*\\* @class \\*/ \\(function \\(_super)', 'g');
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: Some comments explaining what these regular expressions do would be good

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  const importPattern = new RegExp("(?:import)([\\s\\S]*?)(?:';)", 'g');
+  const identifiers: string[] = [];
+  const imports: string[] = [];
+  let temp;
+  let className;
+  const output: ITransformedExample = { output: undefined, error: undefined };
+
+  example = example.replace("import * as React from 'react';", '');
+
+  // Getting classname
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: You should probably go into more detail about what's going on here

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+  while ((temp = importPattern.exec(example))) {
+    if (!/office-ui-fabric-react/.test(temp[0])) {
+      output.error = 'error: unsupported imports';
+    } else {
+      imports.push(temp[0]);
+    }
+  }
+
+  imports.map((imp: string) => {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: Since you're not using the return value, forEach would be more appropriate here.
```suggestion
  imports.forEach((imp: string) => {
```

File: packages/tsx-editor/src/transpiler/transpile.tsx [hunk 20]
```diff
@@ -20,10 +21,15 @@ export async function transpile(model: ITextModel): Promise<ITranspiledOutput> {
   return transpiledOutput;
 }
 
-export function evalCode(code: string): string | undefined {
+export function evalCode(code: string, id: string): string | undefined {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/transpile.tsx]: Would be good to have a doc comment explaining the id param (and probably give it a more descriptive name)

File: packages/utilities/src/ie11Detector.ts [hunk 21]
```diff
@@ -0,0 +1,11 @@
+export const isIE11 = (): boolean => {
+  if (!window || !window.navigator || !window.navigator.userAgent) {
+    return false;
+  }
+
+  if (window.navigator.userAgent.indexOf('rv:11.0') > -1) {
```
Comment (ecraig12345) [inline packages/utilities/src/ie11Detector.ts]: You can just do `return window.navigator.userAgent.indexOf('rv:11.0') > -1;`

File: packages/tsx-editor/src/components/Editor.tsx (hunk 12 above)
Comment (sopranopillow) [reply packages/tsx-editor/src/components/Editor.tsx]: This disables the red lines. I disabled them while nishi works on it.

File: packages/tsx-editor/src/components/EditorContext.tsx (hunk 14 above)
Comment (sopranopillow) [reply packages/tsx-editor/src/components/EditorContext.tsx]: Sorry, forgot to delete the file. It is no longer being used

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 22]
```diff
@@ -37,20 +39,23 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
...
     super(props);
 
     this.state = {
-      isCodeVisible: false,
       schemeIndex: 0,
       themeIndex: 0
     };
+
+    this.canRenderLiveEditor =
+      getSetting('useEditor') === '1' && !isIE11() && transformExample(props.code!, 'placeholder').error === undefined ? true : false;
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Typically we wouldn't include the explicit `? true : false`.
```suggestion
      getSetting('useEditor') === '1' && !isIE11() && transformExample(props.code!, 'placeholder').error === undefined;
```

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 23]
```diff
@@ -121,20 +133,27 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
...
-                </CustomizerContext.Provider>
-              ) : (
-                exampleCardContent
+              {isCodeVisible && !this.canRenderLiveEditor && (
+                <div className={classNames.code}>
+                  <CodeSnippet language="tsx">{code}</CodeSnippet>
+                </div>
               )}
 
+              {isCodeVisible && this.canRenderLiveEditor && editor}
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Instead of declaring the editor JSX up above (which results in always initializing the component even if it's not rendered), you should declare it inline here so it's only initialized when used.

Also, it would probably be better to do the conditional rendering like this:
```
{isCodeVisible && (this.canRenderLiveEditor ? <Editor .../> : /*div with CodeSnippet*/)}
```

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 24]
```diff
@@ -121,20 +133,27 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
...
+
+              {(!isCodeVisible || !this.canRenderLiveEditor) &&
+                (activeCustomizations ? (
+                  <CustomizerContext.Provider value={{ customizations: { settings: {}, scopedSettings: {} } }}>
+                    <Customizer {...activeCustomizations}>
+                      <ThemeProvider scheme={_schemes[schemeIndex]}>
+                        <Stack styles={regionStyles}>{exampleCardContent}</Stack>
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: The customizations need to be applied to the EditorPreview as well, so you'll probably need to do something like this: 
```suggestion
                        <Stack styles={regionStyles}>{this.canRenderLiveEditor ? <EditorPreview ... /> : exampleCardContent}</Stack>
```

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx (hunk 23 above)
Comment (ecraig12345) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: We also need to be sure that Monaco is delay-loaded before this goes in so it doesn't have too much perf impact.

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 25]
```diff
@@ -163,6 +182,27 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
     }
   }
 
+  private _editorOnChange = (editor: ITextModel) => {
+    transpile(editor).then((output: ITranspiledOutput) => {
+      if (output.outputString) {
+        const evalCodeError = evalCode(output.outputString, this.props.title.replace(' ', ''));
+        if (evalCodeError) {
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Could condense this if/else into:
```
this.setState({
  error: evalCodeError || undefined
});
```

File: packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx [hunk 26]
```diff
@@ -27,7 +32,9 @@ export const ExamplesSection: React.StatelessComponent<IExamplesSectionProps> =
             const { view, ...exampleProps } = example;
             return (
               <div key={example.title + '-key'} className={styles.subSection}>
-                <ExampleCard {...exampleProps}>{view}</ExampleCard>
+                <ExampleCard {...exampleProps} onToggleEditor={handleEditorToggle} isCodeVisible={exampleProps.title === activeEditorTitle}>
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/Page/sections/ExamplesSection.tsx]:
```suggestion
                <ExampleCard {...exampleProps} onToggleEditor={setActiveEditorTitle} isCodeVisible={exampleProps.title === activeEditorTitle}>
```

File: packages/utilities/src/ie11Detector.ts [hunk 27]
```diff
@@ -0,0 +1,7 @@
+export const isIE11 = (): boolean => {
+  if (!window || !window.navigator || !window.navigator.userAgent) {
```
Comment (ecraig12345) [inline packages/utilities/src/ie11Detector.ts]:
```suggestion
  if (typeof window === 'undefined' || !window.navigator || !window.navigator.userAgent) {
```

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 28]
```diff
@@ -37,28 +36,36 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
...
     this.state = {
-      isCodeVisible: false,
       schemeIndex: 0,
       themeIndex: 0
     };
+    this.canRenderLiveEditor =
+      getSetting('useEditor') === '1' && !isIE11() && transformExample(props.code!, 'placeholder').error === undefined;
+
+    if (this.canRenderLiveEditor) {
+      import('office-ui-fabric-react').then(Fabric => {
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: I wonder if this should be done in the tsx-editor package instead of here?

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 29]
```diff
@@ -68,17 +75,19 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
...
           const { codeButtons: codeButtonStyles } = subComponentStyles;
 
-          const exampleCardContent = (
-            <div className={classNames.example} data-is-scrollable={isScrollable}>
-              {children}
-            </div>
-          );
+          const exampleCardContent =
+            this.props.isCodeVisible && this.canRenderLiveEditor ? (
+              <EditorPreview id={this.props.title.replace(' ', '')} />
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: You should probably apply `classNames.example` to this too (just add a `className` prop to the component if it doesn't already have one)

Comment (ecraig12345) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Also wonder if it will need `data-is-scrollable` like the normal example div (not sure exactly where that's being used)

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx [hunk 30]
```diff
@@ -162,6 +179,22 @@ export class ExampleCardBase extends React.Component<IExampleCardProps, IExample
       );
     }
   }
+  private _editorOnChange = (editor: ITextModel) => {
+    import('@uifabric/tsx-editor').then(editorModule => {
```
Comment (ecraig12345) [inline packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Probably instead of re-importing this every onChange, you should probably do the import in `_onToggleCodeClick` and save the `editorModule` in a private member.

File: packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx [hunk 31]
```diff
@@ -143,10 +143,10 @@ export class CalendarMonth extends BaseComponent<ICalendarMonthProps, ICalendarM
               {dateTimeFormatter.formatYear(navigatedDate)}
             </div>
           ) : (
-              <div className={css('ms-DatePicker-currentYear js-showYearPicker', styles.currentYear)}>
-                {dateTimeFormatter.formatYear(navigatedDate)}
-              </div>
-            )}
+            <div className={css('ms-DatePicker-currentYear js-showYearPicker', styles.currentYear)}>
```
Comment (ecraig12345) [inline packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx]: Not sure why this change is here?

File: packages/tsx-editor/src/transpiler/exampleTransform.ts (hunk 17 above)
Comment (ecraig12345) [reply packages/tsx-editor/src/transpiler/exampleTransform.ts]: Nice comments! :)

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+
+  /**
+   * Getting all the imports from fabric and pushing them into an array.
+   * If there are imports that do not come from OUFR then it will add an
+   * error since the import is not supported.
+   */
+  while ((temp = importPattern.exec(example))) {
+    if (!/office-ui-fabric-react/.test(temp[0])) {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: Should probably ensure it's at the beginning of the string:
```suggestion
    if (!/^office-ui-fabric-react/.test(temp[0])) {
```

File: packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx (hunk 31 above)
Comment (sopranopillow) [reply packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx]: we didn't do that change, not sure why it's showing up there.

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+   * This will loop through all imports getting and saving the identifiers,
+   * if no identifiers were found then it will just eliminate the import.
+   * If identifiers were found, the string will be split, spaces will be removed,
+   * and next line characters will also be removed.
+   */
+  imports.forEach((imp: string) => {
+    temp = identifierPattern.exec(imp);
+    if (temp !== null) {
+      temp[0].split(',').forEach((ident: string) => {
+        ident.replace('\n', '');
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: `.replace` doesn't mutate the original string, so this won't do anything. Also you could combine this line and the next two as follows (where `\s` is any whitespace character):
```
identifiers.push(ident.replace(/\s/g, ''))
```

File: packages/tsx-editor/src/transpiler/exampleTransform.ts (hunk 32 above)
Comment (sopranopillow) [reply packages/tsx-editor/src/transpiler/exampleTransform.ts]: it shouldn't matter, it just checks if that portion exists in the import.

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+        ident.replace(' ', '');
+        identifiers.push(ident);
+      });
+    }
+    example = example.replace(imp, '');
+  });
+
+  // If there are exports in the code then they will be removed since they are not supported.
+  while (/export/.test(example)) {
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: You should probably test that the `export` is at the beginning of a line and followed by a space. Also you can do this on pass (no while loop needed) like this:
```
example = example.replace(/^export /gm, '');
```

`g` flag means global and `m` means make `^`/`$` match the beginning/end of each line, not just the beginning/end of the whole string.

File: packages/tsx-editor/src/transpiler/transpile.tsx [hunk 35]
```diff
@@ -20,10 +21,24 @@ export async function transpile(model: ITextModel): Promise<ITranspiledOutput> {
   return transpiledOutput;
 }
 
-export function evalCode(code: string): string | undefined {
+/**
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/transpile.tsx]: The formatting/syntax on this comment is a bit off. Also, you don't need the parameter types or the `@export` tag for TS.
```ts
/**
 * Evaluates the code given after transforming it.
 *
 * @param code - Code to transform
 * @param divId - `id` of the `div` element where the example will be rendered into after transforming
 * @returns Returns undefined if the transform was successful, or the error message if it was unsuccessful.
 */
```

Comment (sopranopillow) [reply packages/tsx-editor/src/transpiler/transpile.tsx]: `/**
 * Tranforms the code since the given code might have unsupported imports and then evals the code.
 *
 * @param {code} string
 * @param {divId} string
 * @returns {string | undefined}
 */`

would this sounds better?

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx (hunk 29 above)
Comment (sopranopillow) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: Should I add `data-is-scrollable` just in case?

File: packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx (hunk 31 above)
Comment (ecraig12345) [reply packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx]: Can you try and revert it? Otherwise beachball will think you need a change file for OUFR. You might need to check out the file from master on the command line (instead of manually editing) then commit with the `-n` option to avoid running prettier.
 ```
git checkout master -- packages/office-ui-fabric-react/src/components/Calendar/CalendarMonth.tsx
git commit -a -m "Revert file" -n
```

File: packages/tsx-editor/src/transpiler/exampleTransform.ts [hunk 36]
```diff
@@ -0,0 +1,92 @@
...
+    example = example.replace('export', '');
+  }
+
+  /**
+   * adding line to render React and adding identifiers.
+   */
+  example =
+    'const {' +
+    identifiers.map((identifier: string) => identifier) +
```
Comment (ecraig12345) [inline packages/tsx-editor/src/transpiler/exampleTransform.ts]: You should probably be using join here, not map (also the way you have the map call set up, it's just returning an array identical to the original).
```suggestion
    identifiers.join(', ') +
```

File: packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx (hunk 29 above)
Comment (ecraig12345) [reply packages/example-app-base/src/components/ExampleCard/ExampleCard.tsx]: I think so, though to actually get it applied you might have to add a getNativeProps call in EditorPreview.

File: packages/tsx-editor/src/transpiler/transpile.tsx (hunk 35 above)
Comment (ecraig12345) [reply packages/tsx-editor/src/transpiler/transpile.tsx]: Can you combine the main comment from your suggestion with the params/return part from mine? :)

Comment (sopranopillow) [reply packages/tsx-editor/src/transpiler/transpile.tsx]: sure!

File: apps/fabric-website/webpack.serve.config.js [hunk 37]
```diff
@@ -22,6 +24,14 @@ module.exports = resources.createServeConfig({
     'react-dom': 'ReactDOM'
   },
 
+  plugins: [
+    new MonacoWebpackPlugin({
+      // available options are documented at https://github.com/Microsoft/monaco-editor-webpack-plugin#options
+      languages: ['typescript']
+    }),
+    new BundleAnalyzerPlugin()
```
Comment (ecraig12345) [inline apps/fabric-website/webpack.serve.config.js]: Remove this now that we're done testing

**Comment by sopranopillow** [comment]
@ecraig12345 it appears it's failing when running tests for some component that we didn't modify. Is this happening to other PRs?
